src/probnum/_config.py

```diff
@@ -0,0 +1,91 @@
+    Additionally, the user can register own configuration entries via the method
+    :meth:`register`. Configuration entries can only be registered once and can only
+    be used (accessed or overwritten) once they have been registered.
+
+    +----------------------------------+-----------------------------------------------+
+    | Config entry                     | Description                                   |
+    +==================================+===============================================+
+    | ``covariance_inversion_damping`` | A (typically small) value that is per         |
+    |                                  | default added to the diagonal of covariance   |
+    |                                  | matrices in order to make inversion           |
+    |                                  | numerically stable.                           |
+    +----------------------------------+-----------------------------------------------+
+    | ``...``                          | ...                                           |
+    +----------------------------------+-----------------------------------------------+
+
+    Examples
+    ========
+
+    >>> import probnum
+    >>> probnum.config.covariance_inversion_damping
+    1e-12
+    >>> with probnum.config(
+    ...     covariance_inversion_damping=1e-2,
+    ... ):
+    ...     probnum.config.covariance_inversion_damping
+    0.01
+    """
+
+    @contextlib.contextmanager
+    def __call__(self, **kwargs) -> None:
+        old_entries = dict()
+
+        for key, value in kwargs.items():
+            if not hasattr(self, key):
+                raise KeyError(
+                    f"Configuration entry {key} does not exist yet."
+                    "Configuration entries must be `register`ed before they can be "
+                    "accessed."
+                )
+
+            old_entries[key] = getattr(self, key)
+
+            setattr(self, key, value)
+
+        try:
+            yield
+        finally:
+            self.__dict__.update(old_entries)
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        if not hasattr(self, key):
+            raise KeyError(
+                f"Configuration entry {key} does not exist yet."
+                "Configuration entries must be `register`ed before they can be "
+                "accessed."
+            )
+
+        self.__dict__[key] = value
+
+    def register(self, key: str, default_value: Any) -> None:
+        if hasattr(self, key):
+            raise KeyError(
+                f"Configuration entry {key} does already exist and "
+                "cannot be registered again."
+            )
+        self.__dict__[key] = default_value
+
+
+# Create a single, global configuration object,...
+_GLOBAL_CONFIG_SINGLETON = Configuration()
+
+# ... define some configuration options, and the respective default values
+# (which have to be documented in the Configuration-class doctstring!!), ...
```

Comment:
```suggestion
# (which have to be documented in the Configuration-class docstring!!), ...
```

Reply:
I assume there is no good way of documenting these options here automatically (including their default values), is there? Manually updating entries and their value in the documentation is prone to breaking frequently.

Reply:
> I assume there is no good way of documenting these options here automatically (including their default values), is there?

Not that I know of.

> Manually updating entries and their value in the documentation is prone to breaking frequently.

Agreed 100%. Still - in my opinion - better than a half-baked automated solution.

Reply:
Then I suggest to manually add the default values to the table in the docstring of `Configuration`. That way they are exposed to the user.